### PR TITLE
[AutoDiff] Plumb witness derivative generic signatures through SILGen.

### PR DESCRIPTION
**Pull request by rxwei**
When SILGenWitnessTable creates a decl ref for the witness of a derivative function requirement, it is using the requirement's derivative generic signature in the resulting witness decl ref. This is wrong because the witness may have a different derivative generic signature than the requirement, leading to a crash. This bug was never discovered because GSB's dark magic made it "just work", until requirement machine.

The fix is to store the matched witness derivative generic signature in `Witness` during type checking, and during witness table generation, use the witness' generic signature to create a witness decl ref.

Also brings back #39505 by reverting the revert (#40057), now that the blocking issue was fixed.

Resolves rdar://84716758, rdar://84213107 and rdar://84987079.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test

**Comment by rxwei** [comment]
Since this PR contains a revert-of-revert, see https://github.com/apple/swift/pull/40063/commits/ca4a645a065190526721dc42f2857eea4320db6d for the diff specific to this fix.

File: test/AutoDiff/SILGen/vtable.swift [hunk 1]
```diff
@@ -1,4 +1,6 @@
-// RUN: %target-swift-frontend -emit-silgen %s -requirement-machine=off | %FileCheck %s
+// RUN: %target-swift-frontend -emit-silgen %s | %FileCheck %s
+
+// FIXME(rdar://84987079)
```
Comment (slavapestov) [inline test/AutoDiff/SILGen/vtable.swift:3]: What is this FIXME in relation to? The test appears to be enabled

**Review by slavapestov — APPROVED** [review]
Thanks!

**Comment by rxwei** [comment]
@swift-ci please smoke test

File: test/AutoDiff/SILGen/vtable.swift (hunk 1 above)
Comment (rxwei) [reply test/AutoDiff/SILGen/vtable.swift]: This was accidental. I deleted it.

**Comment by rxwei** [comment]
@swift-ci please smoke test macos

**Pull request merged**